tests/ashley/test_forum_search.py

```diff
@@ -0,0 +1,401 @@
+        user = UserFactory()
+
+        # Index the post in Elasticsearch
+        call_command("rebuild_index", interactive=False)
+
+        self.client.force_login(user)
+        response = self.client.get("/forum/search/?q=*")
+        self.assertContains(
+            response, "Your search has returned <b>0</b> results", html=True
+        )
+
+    def test_forum_search_anonymous(self):
+        """Anonymous users should not get any results."""
+        PostFactory()
+
+        # Index the post in Elasticsearch
+        call_command("rebuild_index", interactive=False)
+
+        response = self.client.get("/forum/search/?q=*")
+        self.assertContains(
+            response, "Your search has returned <b>0</b> results", html=True
+        )
+
+    def test_forum_search_post_content_partial(self):
+        """Searching for part of a word contained in the post content should find it."""
+        post = PostFactory(text="a5g3g6k75")
+
+        user = UserFactory()
+        assign_perm("can_read_forum", user, post.topic.forum)
+
+        # Index the post in Elasticsearch
+        call_command("rebuild_index", interactive=False)
+
+        self.client.force_login(user)
+        response = self.client.get("/forum/search/?q=a5g3")
+        self.assertContains(
+            response, "Your search has returned <b>1</b> result", html=True
+        )
+        self.assertContains(response, post, html=True)
+
+    def test_forum_search_post_subject_partial(self):
+        """Searching for part of a word contained in the post subject should find it."""
+        post = PostFactory(subject="f61f59ky2")
+
+        user = UserFactory()
+        assign_perm("can_read_forum", user, post.topic.forum)
+
+        # Index the post in Elasticsearch
+        call_command("rebuild_index", interactive=False)
+
+        self.client.force_login(user)
+        response = self.client.get("/forum/search/?q=f61f")
+        self.assertContains(
+            response, "Your search has returned <b>1</b> result", html=True
+        )
+        self.assertContains(response, post.subject, html=True)
+
+    def test_forum_search_post_poster_partial_public_username(self):
+        """Searching for part of the poster public username should find the post."""
+        post = PostFactory(poster__public_username="3vj9tq3gc")
+        user = UserFactory()
+        assign_perm("can_read_forum", user, post.topic.forum)
+
+        # Index the post in Elasticsearch
+        call_command("rebuild_index", interactive=False)
+
+        self.client.force_login(user)
+        response = self.client.get("/forum/search/?q=&search_poster_name=3vj9t")
+        self.assertContains(
+            response, "Your search has returned <b>1</b> result", html=True
+        )
+        self.assertContains(response, post.subject)
+
+    def test_forum_search_post_topic_partial(self):
+        """
+        The topic subject is not directly indexed but Django Machina forces it with the post
```

Comment:
Does it mean that we have to run the `rebuild_index` command on a regular basis in order to index new posts ?

Reply:
Ok, in fact the answer to my question is documented in both `django-haystack` and `django-machina`.

In [Django Haystack](https://django-haystack.readthedocs.io/en/master/tutorial.html#reindex) : 

> Using the standard SearchIndex, your search index content is only updated whenever you run either ./manage.py update_index or start afresh with ./manage.py rebuild_index.
>
> You should cron up a ./manage.py update_index job at whatever interval works best for your site (using --age=<num_hours> reduces the number of things to update).
>
> Alternatively, if you have low traffic and/or your search engine can handle it, the RealtimeSignalProcessor automatically handles updates/deletes for you.

In [Django machina](https://django-machina.readthedocs.io/en/stable/getting_started.html#django-haystack-settings) : 

> It should be noticed that you’ll have to run the update_index Haystack’s command once your forum is properly set up in order to make your topics & posts searchable.


Django machina does not implement any realtime index update with signals, since it can be done in several ways ways with haystack, depending on the search backend we use, the traffic of the website and the performance impact. 

I think we can merge it as is as a first step and set up a cronjob to update the index on a regular basis.

Then, we can consider a more «realtime» solution, based on [RealtimeSignalProcessor](https://django-haystack.readthedocs.io/en/master/signal_processors.html#realtime-realtimesignalprocessor), a custom Processor, or with a job queue (see [these apps](https://django-haystack.readthedocs.io/en/master/other_apps.html#ref-other-apps))

Reply:
Nice.
I think in our case (big index, few updates) the realtime update of any record that changes is the best.

Reply:
I agree (to only update the record that change). And doing this in a celery task would be better!

Reply:
Actually, it is not more costly than a write to postgresql... and writes are rare on a forum? If adding Celery to the stack can be avoided...

Reply:
I said this after reading Haystack's [RealtimeSignalProcessor](https://django-haystack.readthedocs.io/en/master/signal_processors.html#realtime-realtimesignalprocessor) documentation : 

> These updates happen in-process, which if a request-response cycle is involved, may cause the user with the browser to sit & wait for indexing to be completed. Since this wait can be undesirable, especially under load, you may wish to look into queued search options. See the Haystack-Related Applications documentation for existing options.

I don't know the cost of a write to Elasticsearch and what is doing the RealtimeSignalProcessor exactly yet.
May I suggest to think about this evolution and discuss it in a second time ?

Lets start with a simple cronjob :-)